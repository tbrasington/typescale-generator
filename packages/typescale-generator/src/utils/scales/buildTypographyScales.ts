import { ViewPortProps, GeneratedNamedScalesProps, stepValues } from ".";
import { round } from "../round";
import { clampBuilder } from "./clampBuilder";

/**
 * @returns An array of units and clamped values.
 * ```js
 *    {
 *      typeScale = {
 *         "0": {
 *           clamp : 'clamp(1rem, 0.93rem + 0.36vw, 1.25rem)',
 *           min : {
 *              fontSize: 14.05,
 *              width: 320,
 *              scale : 1.25
 *          },
 *          max : {
 *              fontSize: 17.57,
 *              width: 1440,
 *              scale : 1.25
 *          }
 *        }
 *      }
 *    }
 * ```
 * @example 
 * ```js
*   buildTypographyScales({
      min: {
        width: 320,
        fontSize: 16,
        typeScale: TypographyScaleValues[scale].value,
      },
      max: {
        width: 1440,
        fontSize: 20,
        typeScale: TypographyScaleValues[scale].value,
      },
     range: [-2, -1, 0, 1, 2, 3, 4, 5],
  });
  ```
 */
export function buildTypographyScales(opts: {
  range: number[];
  min: ViewPortProps;
  max?: ViewPortProps;
}) {
  const { min, max, range } = opts;

  const system: GeneratedNamedScalesProps = {
    typeScale: [],
  };

  const minFluidTypeStep = range[0];
  const maxFluidTypeStep = range[range.length - 1];
  for (let i = minFluidTypeStep; i <= maxFluidTypeStep; i++) {
   
    system.typeScale[i] = stepValues({
      min: min,
      max: max,
      step: i,
    })
  }

  return {
    ...system,
  };
}