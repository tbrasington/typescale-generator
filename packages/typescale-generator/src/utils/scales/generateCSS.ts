import { GeneratedItemProps, GenericKeyValueProps, TypeScaleProps } from "../interfaces";


export function generateCSS({ scales }: { scales: TypeScaleProps; }) {
  const typeSteps: GenericKeyValueProps = {};

  for (const step of Object.keys(scales)) {
    typeSteps[`--${step}`] = `${scales[`${step}`].clamp};`;
  }

  return `:root ${JSON.stringify(typeSteps, null, "  ").replace(/"/g, "").replace(/,/g, "")}`;
}
