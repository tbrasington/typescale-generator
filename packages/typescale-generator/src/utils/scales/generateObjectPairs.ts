import { GenericKeyValueProps, TypeScaleProps } from "../interfaces";

/**
 * @returns An object of the steps with the min and max unit values, paired with the maxwidth.
 * @example ```js
 *
 * ```
 */

export function generateObjectPairs({ scales }: { scales: TypeScaleProps }) {
   
  const remap = Object.keys(scales).map((item) => {
   const isMaxAvailable = (scales[item].max?.width !== undefined) ? true : false;

    return {
      step: item,
      fontSize: {
        [scales[item].min.width]: scales[item].min.fontSize,
        //@ts-ignore
        ...(isMaxAvailable && {[scales[item].max?.width] : scales[item].max?.fontSize})
      },
    };
  });

  return remap;
}
