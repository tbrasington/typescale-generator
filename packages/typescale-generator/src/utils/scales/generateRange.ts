import { RangeProps } from ".";

/** 
* 
* Generates a range of numbers to use as the steps in the typography scale
* @returns [-2, -1, 0, 1, 2, 3, 4, 5]
*/
export function generateRange({min, max} : RangeProps)  {
  return Array.from(new Array(max + 1 + -1 * min), (_x, i) => i + min);
}
