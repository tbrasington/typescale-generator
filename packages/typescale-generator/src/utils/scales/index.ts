export { generateRange } from "./generateRange";
export { stepValues } from "./stepValues";
export { numberToRem } from "./numberToRem";
export { clampBuilder } from "./clampBuilder";
export { buildTypographyScales } from "./buildTypographyScales";
export { generateObjectPairs } from "./generateObjectPairs";
export { generateCSS } from './generateCSS'
export { generateObject } from './generateObject'