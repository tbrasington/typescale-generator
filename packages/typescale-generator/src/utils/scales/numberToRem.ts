export function numberToRem(px: number): string {
  return `${px / 16}rem`;
}