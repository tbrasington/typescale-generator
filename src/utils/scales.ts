

interface viewPortProps {
  width: number;
  fontSize: number;
  typeScale: number;
}

interface SystemProps {
  typeScale: systemItemProps[];
}

interface systemItemProps {
  min: number;
  max: number;
  clamp: string;
}

export const SCALES = {
  MAJOR_SECOND: "MAJOR_SECOND",
  MINOR_SECOND: "MINOR_SECOND",
  MAJOR_THIRD: "MAJOR_THIRD",
  MINOR_THIRD: "MINOR_THIRD",
  PERFECT_FOURTH: "PERFECT_FOURTH",
  AUGMENTED_FOURTH: "AUGMENTED_FOURTH",
  PERFECT_FIFTH: "PERFECT_FIFTH",
  GOLDEN_RATIO: "GOLDEN_RATIO",
  OCTAVE: "OCTAVE",
  MAJOR_SIXTH: "MAJOR_SIXTH",
  MINOR_SEVENTH: "MINOR_SEVENTH",
  MAJOR_SEVENTH: "MAJOR_SEVENTH",
};

export const TypographyScaleValues = {
  [SCALES.MINOR_SECOND]: { name: "Minor second", value: 1.067 },
  [SCALES.MAJOR_SECOND]: { name: "Major second", value: 1.125 },
  [SCALES.MINOR_THIRD]: { name: "Minor third", value: 1.2 },
  [SCALES.MAJOR_THIRD]: { name: "Major third", value: 1.25 },
  [SCALES.PERFECT_FOURTH]: { name: "Perfect fourth", value: 1.333 },
  [SCALES.AUGMENTED_FOURTH]: {
    name: "Augmented fourth",
    value: 1.414,
  },
  [SCALES.PERFECT_FIFTH]: { name: "Perfect fifth", value: 1.5 },
  [SCALES.GOLDEN_RATIO]: { name: "Golden ratio", value: 1.618 },
  [SCALES.MAJOR_SIXTH]: { name: "Major sixth", value: 1.667 },
  [SCALES.MINOR_SEVENTH]: { name: "Minor seventh", value: 1.778 },
  [SCALES.MAJOR_SEVENTH]: { name: "Major seventh", value: 1.875 },
  [SCALES.OCTAVE]: { name: "Octave", value: 2 },
};

function round(num: number) {
  return Math.round(num * 100) / 100;
}

function clampBuilder(
  minWidthPx: number,
  maxWidthPx: number,
  minSizePx: number,
  maxSizePx: number
) {
  const pixelsPerRem = 16;

  const minWidth = minWidthPx / pixelsPerRem;
  const maxWidth = maxWidthPx / pixelsPerRem;

  const minSize = minSizePx / pixelsPerRem;
  const maxSize = maxSizePx / pixelsPerRem;

  const slope = (maxSize - minSize) / (maxWidth - minWidth);
  const yAxisIntersection = -minWidth * slope + minSize;

  const min = `${round(minSize)}rem`;
  const val = `${round(yAxisIntersection)}rem + ${round(slope * 100)}vw`;
  const max = `${round(maxSize)}rem`;

  return `clamp(${min}, ${val}, ${max})`;
}

export function generateCSS(system: { typeScale: any }) {
  let css = "";

  const { typeScale } = system;

  for (const step of Object.keys(typeScale)) {
    css += `--step-${step}: ${typeScale[step].clamp};`;
  }
  return css;
}


interface cssProps { 
  [key: string]: string 
}

export function generateTypeScript({ scales}  : {scales: systemItemProps[]}) {


const typeSteps : cssProps = {} 

  for (const step of Object.keys(scales)) {
    typeSteps[`step-${step}`] = scales[parseInt(step)].clamp
  }

  return typeSteps
}
export function buildFluidDesignSystem(opts: {
  typeScaleSteps: number[];
  minViewport: viewPortProps;
  maxViewport: viewPortProps;
}) {
  const { minViewport, maxViewport, typeScaleSteps } = opts;

  const system: SystemProps = {
    typeScale: [],
  };

  const minFluidTypeStep = typeScaleSteps[0];
  const maxFluidTypeStep = typeScaleSteps[typeScaleSteps.length - 1];

  for (let i = minFluidTypeStep; i <= maxFluidTypeStep; i++) {
    const valueMin = round(
      minViewport.fontSize * Math.pow(minViewport.typeScale, i)
    );

    const valueMax = round(
      maxViewport.fontSize * Math.pow(maxViewport.typeScale, i)
    );

    system.typeScale[i] = {
      min: valueMin,
      max: valueMax,
      clamp: clampBuilder(
        minViewport.width,
        maxViewport.width,
        valueMin,
        valueMax
      ),
    };
  }

  return {
    ...system,
  };
}

export function generateScales() {
  const config = {
    minViewport: {
      width: 320,
      fontSize: 16,
      typeScale: TypographyScaleValues[SCALES.MINOR_SECOND].value,
    },
    maxViewport: {
      width: 1440,
      fontSize: 20,
      typeScale: TypographyScaleValues[SCALES.MINOR_SECOND].value,
    },
  };

  const fluidDesignSystem = buildFluidDesignSystem({
    ...config,
    typeScaleSteps: [-2, -1, 0, 1, 2, 3, 4, 5],
  });

  return fluidDesignSystem;
}
