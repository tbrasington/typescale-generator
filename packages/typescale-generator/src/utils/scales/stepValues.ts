import { ViewPortProps } from "../interfaces";
import { round } from "../round";
import { clampBuilder } from "./clampBuilder";
import { numberToRem } from "./numberToRem";


/**
 * @returns An object of CSS clamped values and their min and max number values, corresponding to their step.
 * ```js
 *    { 
 *       clamp : 'clamp(1rem, 0.93rem + 0.36vw, 1.25rem)',
 *       min : {
 *        fontSize: 14.05,
 *         width: 320,
 *         scale : 1.25
 *          },
 *          max : {
 *              fontSize: 17.57,
 *              width: 1440,
 *              scale : 1.25
 *          }
 *        }
 *      
 * ```
 * @example 
 * ```js
*   stepValues({
      min: {
        width: 320,
        fontSize: 16,
        typeScale: TypographyScaleValues[scale].value,
      },
      max: {
        width: 1440,
        fontSize: 20,
        typeScale: TypographyScaleValues[scale].value,
      },
      step : 1
  });
  ```
 */
export function stepValues({min, max, step}: {
  min: ViewPortProps;
  max?: ViewPortProps;
  step: number;
}) {
    const valueMin = round(
      min.fontSize * Math.pow(min.typeScale, step)
    );

    const valueMax = max ? round(
      max.fontSize * Math.pow(max.typeScale, step)
    ) : undefined;

    return {
      min: {
        fontSize: valueMin,
        width: min.width,
        typeScale: min.typeScale,
      },
      max: valueMax && max ? {
        fontSize: valueMax,
        width: max.width,
        typeScale: max.typeScale,
      } : undefined,
      clamp: valueMax && max ? clampBuilder(
        min.width,
        max.width,
        valueMin,
        valueMax
      ) : numberToRem(valueMin),
    }
}