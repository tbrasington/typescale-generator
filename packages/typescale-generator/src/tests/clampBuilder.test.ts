import { clampBuilder } from "../utils/scales/clampBuilder"


  test('Generate clamped sizes for css', () => {
    const clamped = clampBuilder(320,1440,14.05,17.57)
    expect(clamped).toMatchSnapshot()
  })
 