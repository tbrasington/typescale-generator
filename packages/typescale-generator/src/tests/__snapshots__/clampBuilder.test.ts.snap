// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate clamped sizes for css 1`] = `"clamp(0.88rem, 0.82rem + 0.31vw, 1.1rem)"`;
